include/circt/Dialect/FIRRTL/FIRRTLTypes.td

```diff
@@ -54,6 +54,63 @@ def PassiveType : DialectType<FIRRTLDialect,
   CPred<"$_self.isa<FIRRTLType>() && $_self.cast<FIRRTLType>().isPassive()">,
   "a passive type (contain no flips)", "::circt::firrtl::FIRRTLType">;
 
+def CMemoryType : TypeDef<FIRRTLDialect, "CMemory"> {
+  let mnemonic = "cmemory";
+  let summary = "a behavioral memory";
+  let description = [{
+    Syntax:
+    ```
+    cmemory-type ::= `cmemory` `<` element-type, element-count `>`
+    ```
+
+    The value of a `cmemory` type represents a behavioral memory with unknown
+    ports.  This is produced by `combmem` and `seqmem` declarations and used by
+    `memoryport` declarations to define memories and their ports. A CMemory is
+    similar to a vector of passive element types.
+
+    Examples:
+
+    ```mlir
+    !firrtl.cmemory<uint<32>, 16>
+    !firrtl.cmemory<bundle<a : uint<1>>, 16>
+    ```
+  }];
+
+  let parameters = (ins "FIRRTLType":$elementType, "unsigned":$numElements);
+
+  let builders = [
+    TypeBuilderWithInferredContext<(ins "FIRRTLType":$elementType,
```

Comment:
Plz move the parsing/printing logic out of line, and use the `||` combinator in the parser logic to collapse the if's